doc/dev/architecture.md

```diff
@@ -0,0 +1,223 @@
+    *adjoint-jacobian* method, a technique similar to classical backpropagation. By taking
+    advantage of the reversibility of quantum computing, a backwards pass can be performed with a
+    much lower memory footprint than with backpropagation. Hardware compatible methods can directly
+    be applied in the compiler without requiring explicit device support. This includes the
+    *parameter-shift* method and *finite-differences*. The parameter-shift method has been adapted
+    to work in presence of hybrid program representations including control flow, as long as
+    measurement feedback is not used.
+
+  - Checkpointing is employed to eliminate redundant computations of the pre-processing, by storing
+    interemediate results and control flow information in a forwards pass through the classical code
+    to allow the quantum program to be reconstructed exactly.
+
+- **Classical optimizations:**
+
+  - Basic optimizations are frequently performed in between other passes in order to improve
+    performance and reduce the computational load of subsequent transformations. This includes
+    dead code elimation, common sub-expression elimination, and constant propagation, as well
+    other simplifactions (cannonicalizations) registered to the various dialect operations.
+
+  - More advanced optimzation techniques might also be added to various parts of the pass pipeline.
+
+- **Bufferization:**
+
+  - Bufferization is a process by which operations are transformed from operating on *tensors* to
+    operating on memory, represented by *memrefs* (Memory References) in MLIR. The key difference
+    between the two is that tensors behave according to value semantics, that is they cannot be
+    modified in-place. Instead, operations consume and produce new tensor values.
+
+  - In order to bufferize a program, memory has to be allocated and a buffer assigned to each
+    tensor, reusing buffers whenever possible to minimize unnecessary data copies, and eventually
+    deallocating buffers when they are no longer needed to prevent memory leaks.
+
+  - Bufferization should generally be the last step before converting to the LLVM dialect, as
+    optimizations are typically easier to implement in the tensor domain than in the memory domain.
+
+- **LLVM dialect generation:**
+
+  - As an intermediate step in the LLVM IR generation, the LLVM dialect in MLIR provides a simple
+    target for other dialects to lower to, and simplifies the conversion between MLIR and LLVM IR
+    by providing a one-to-one mapping from MLIR objects to LLVM IR objects.
+
+  - Generally, dialects will make use of the dialect conversion infrastructure to provide lowerings
+    to the LLVM dialect. The quantum dialect provides lowering patterns to QIR-style operations, and
+    the Catalyst gradient dialect can lower to device-based implementations of quantum AD or to an
+    Enzyme-based implementation for AD on classical code.
```

Comment:
Hmm, I wanted abstain from listing explicit passes for example because that could easily change on regular basis. Similarly, I could imagine the code structure changing as well invalidating links we include here. 🤔 
But if you think it's still a good idea I can include a link.

Reply:
If things are still subject to a large degree of arch changes, then hold off. If things will be stable for 6-12 months, should be fine to add some links here.

Reply:
Got it, I think I will hold off for now 👍